### PR TITLE
Refactor/rewrite as middleware

### PR DESCRIPTION
**Pull request by PAkerstrand**
This PR re-writes `react-router-redial` to work as a middleware. This allows it to be easily integrated with other middlewares, such as `react-router-scroll`.
## Notable changes
- `useRedial` - New function for creating a React Router middleware. Takes the same options as the earlier `RedialContext`. Instead of calling `<Router render={(props) => <RedialContext .../>}/>`, we now use `applyRouterMiddleware` instead: `<Router render={applyRouterMiddleware(useRedial(redialOptions))}/>`
- The redial-map is now based on `path` from `Route` rather than the `Component`. This allows us to have multiple routes handled by the same component but with different props. On the other hand, if we want to retain the properties, this is no longer possible. **Perhaps we should have a way to force the redial-props key?**
## Remaining work
- [x] Right now the `prevProps` etc doesn't work any longer. How should we handle this? Move the state down to `<RedialContextContainer/>` instead?
- [x] Update the documentation to highlight the new approach to render both on server and client.
- [x] Create an example mixing `useRedial` with another router middleware
  
  ``` jsx
  render={(props) => (
    <RedialContext {...props} {...redialOptions}>
      <RouterContext {...props} 
          createElement={(Component, routeProps) => (
             <RedialContextContainer routerProps={routeProps}>
               <Component/>
             </RedialContextContainer>
          )
      />
   </RedialContext>
  )}
  ```
- [x] Add tests for the util functions. Also, is there a way to gracefully fail rather than throwing?
- [ ] ... More?

### CODE REVIEW

File: src/RedialContext.js [hunk 1]
```diff
@@ -129,7 +122,7 @@ export default class RedialContext extends Component {
   }
 
   reloadComponent(component) {
-    this.load(component, this.props, true);
+    this.load([component], this.props, true);
```
Comment (dlmr) [inline src/RedialContext.js]: I do not think we need to do this here, `redial` should take care of this for us.

Comment (PAkerstrand) [reply src/RedialContext.js]: Reverted

**Comment by dlmr** [comment]
Thanks a ton, this looks awesome! 👍 

Just a quick thought, what do you think about using camelCase for the new util files.

```
get-route-path.js => getRoutePath.js
find-route-by-component.js => findRouteByComponent.js
map-keys.js => createMapKeys.js
```

I will take a closer look at this tomorrow and play around with it a little.

**Comment by PAkerstrand** [comment]
Oh absolutely, just an old habit I guess. Renaming would make it more consistent!

**Comment by dlmr** [comment]
This works great, awesomely done once again! 👍 

I will go ahead and merge this and release it as `0.3.0` soon!

I will add you as contributor in the `README.md` and `package.json` if you don't mind along with giving you write access to the repo.

**Pull request merged**